articles/102_dependency_groups.md

```diff
@@ -0,0 +1,164 @@
+
+### (1) Need at least one "Provider"
+
+A package might want to declare that at least one dependency from a group needs to be present (called `satisfy-at-least-one`).
+Compared to "normal" dependencies it is not required that all dependencies of that group need to be present to satisfy the dependency group.
+
+An example for this is the `rmw_implementation` package.
+Currently it depends on all RMW implementation packages (e.g. `rmw_fastrtps_cpp`, `rmw_connext_cpp`, etc.) in order to ensure that it may only be built after all RMW implementation packages have been built.
+The goal would be that at least one RMW implementation needs to be present in order to build or use the `rmw_implementation` package.
+It would be acceptable if more than one or even all RMW implementations are available but that is not required.
+
+In case multiple package declare to be part of that dependency group the `rmw_implementation` package might want to provide an ordered list of preferred names which should be considered in that order.
+
+### (2) All packages of a specific "type"
+
+Another use case is that a dependency group doesn't need to be explicitly defined by all dependency names (called `satisfy-all`).
+Instead a dependency group is defined by an identifier and the expectation is that all packages which declare that they are being part of that group should be available.
+
+An example for this is the `ros1_bridge` package.
+Currently it depends on a manually selected set of message packages.
+The goal would be that it can declare that it depends on *all* packages which provide messages.
+
+Since the `ros1_bridge` manifest only contains the name of the group dependency, it doesn't control which packages declare that they are part of that group.
+It might want to exclude specific packages to be considered even though they are in the group.
+
+## Processes
+
+The following two processes need to be update to consider dependency groups.
+
+### Build tool
+
+The build tool needs to consider the declared dependency groups when computing the topological order.
+Since it operates on a set packages in a workspace it has all manifests for all of these packages available.
+Independent of the semantic of the dependency group (satisfy-all vs. satisfy-at-least-one) the build tool can simply expand a dependency group to all packages of the set.
+
+### Release tool
+
+Since the release tool `bloom` operates on a single package it can't resolve a dependency group without further information.
+It will need all package manifests available from `rosdistro` to resolve a dependency group to a set of packages.
+
+For the `satisfy-all` semantic it can simply generate dependencies on all packages from the group.
+For the `satisfy-at-least-one` semantic the release tool could either use a platform specific mechanism like `provides` in Debian control files or simply expand the group to all packages (same as `satisfy-all`).
```

Comment:
I see a concern with using the `satisfy` semantic to determine whether or not the release tool uses a platform packaging mechanic like `provides`: From my read of the document thus far, there is no distinction in the group membership declaration between a group which is meant to be used with `satisfy-at-least-one` (Use case (1)) and a group intended for use with `satisfy-all` (Use case (2))

The `Provides` mechanism is used on the group member packages, not on the depending packages which is where the distinction between `satisfy-all` and `satisfy-at-least-one` is made.

Groups intended for both use case (1) and (2) are not distinguished and would by extension share the same release template in bloom (or any other per-package release tool) so virtual packages generated by the use of `Provides` in Debian would be created for both groups of rmw implementations and groups of msg containing packages.

I think this is only a serious problem if creating virtual packages for use case (2) groups is undesirable rather than superfluous.

The Debian policy manual has [more info](https://www.debian.org/doc/debian-policy/#virtual-packages-provides) on how virtual packages and provides works. I notice that virtual packages cannot accept version constraints nor can a package provide a specific version of a virtual package. That may or may not be important down the line. It might be worth looking at other package specifications group and provider mechanics, particularly RPM to make sure that overlapping these two concerns will work on other platforms too.

Reply:
> From my read of the document thus far, there is no distinction in the group membership declaration between a group which is meant to be used with `satisfy-at-least-one` (Use case (1)) and a group intended for use with `satisfy-all` (Use case (2))

The group dependency will have an attribute which identifies which `satisfy` semantic is desired: see example https://github.com/ros2/design/pull/138/files#diff-c8d974466de0d80b2299df802e7b1d7bR132

Reply:
The dependent package specifies the desired semantic but that is unknown the to the group members themselves: https://github.com/ros2/design/blame/0534486af1398a3410159130183ed4c918dcdbbd/articles/102_dependency_groups.md#L141

Also, one package may have `<group_depend satisfy="all">i_am_a_rmw_implementation</group_depend>` while another may depend on the same group with `satisfy-at-least-one`:  `<group_depend satisfy="at-least-one">i_am_a_rmw_implementation</group_depend>`

Reply:
Assuming `bloom` decides to generate a `Provides` entry for the `member_of_group` tag using the name of the dependency name. In that case the package declaring the `group_depend` tag will still use the `satisfy` attribute to either expand all group members (in case of `all`) or depend on the group name (in case of `at-least-one`) which will then rely on the `Provides` to be present.

I don't see a problem with this since the package which needs to make the semantic decision does have the `satisfy` attribute. The members of the group simply do always the same thing (generate `Provides` with the group name, if anybody uses it with `at-least-one` or not is not relevant) since they are not aware of the way they are being used.

Maybe I am missing your point?

Reply:
> The members of the group simply do always the same thing

Ah, because the mention of the provides behavior was only in the paragraph about `any/at-least-one` it makes it seem like the release tool needed to know the difference. Perhaps the section about group member behavior should be pulled above of the section about dependency resolution for each use case.

This brings up a secondary concern about whether this overlapping group mechanic will be suitable for packaging systems other than Debian. From my reading of the debian documentation there's no ill effects from a package providing a virtual package that isn't used. I mentioned previously that I'd take a look at Homebrew's mechanisms for this and will do so first thing tomorrow.